### DISCUSSION TITLE
Report missing tests for public functions

### DISCUSSION THREAD

**Discussion by grrinchas** [jfmengels/elm-review#102] [Rule ideas]
**Dependent rules:**

This rule would not work if rule https://github.com/jfmengels/elm-review/discussions/101 is not enabled. 

**What problems does it solve:**

I had this problem when you create a module, then write tests for that module, and then you forget about that module. Later you decide to add more functions to that module, but you either forget or are just lazy to add corresponding tests. Therefore, you end up with less test coverage. 

**Example of things the rule would report:**

In the following example we have `User.elm` model which exposes two functions `encode` and `toString`. Then we have corresponding `UserTest.elm` module, which has only `toStringTest` function. The rule would detect that `encodeTest` is missing. It would not have any effect on the `decode` function, cause it is private.

```elm
module User exposing (encode, toString)

toString: User -> String
toString user = ...

encode: User -> Encode.Value
encode user = ...

decode: Decoder User
decode = ...
```

```elm
module UserTest exposing (toStringTest)

toStringTest: Test
toStringTest = ...
```

**Example of things the rule would not report:**

* It would not report if the private functions does not have tests
* It would not report any modules which have no tests at all

**When to enable this rule:**

This rule is not a replacement for a proper test coverage tool, but it does have one advantage is that it reports only those modules which have already been tested. Therefore you would enable it if

* You have lots of untested modules and wish to increase little by little test coverage
* You want to keep up to date your already tested modules 
* You are using your tests as requirements documentation
* You want to write better unit tests, cause this rule make it really obvious which test supposed to test which part of the code.

**When (not) to enable this rule:**

* If you do strict TDD, then this rule becomes irrelevant
* You are using test coverage tool

**I am looking for:**

I have already written this rule and been using for a while. Wondering if I should publish it or not? And if publish, whether under my own name or more official one?

**Comment by sparksp** [comment]
I'm less excited by this rule because I rarely have just one test per exposed function. Most of my test modules expose an `all` function which describes and calls the various function tests, which usually list the various cases.
